homeassistant/components/insteon/__init__.py

```diff
@@ -24,21 +25,71 @@
     CONF_SUBCAT,
     CONF_UNITCODE,
     CONF_X10,
-    CONF_X10_ALL_LIGHTS_OFF,
-    CONF_X10_ALL_LIGHTS_ON,
-    CONF_X10_ALL_UNITS_OFF,
     DOMAIN,
-    INSTEON_ENTITIES,
+    INSTEON_COMPONENTS,
+    ON_OFF_EVENTS,
 )
 from .schemas import CONFIG_SCHEMA  # noqa F440
-from .utils import async_register_services, register_new_device_callback
+from .utils import (
+    add_on_off_event_device,
+    async_register_services,
+    get_device_platforms,
+    register_new_device_callback,
+)
 
 _LOGGER = logging.getLogger(__name__)
 
 
+async def async_id_unknown_devices(config_dir):
+    """Send device ID commands to all unidentified devices."""
+    _LOGGER.info("Identifying Insteon devices")
+    await devices.async_load(id_devices=1)
+    for addr in devices:
+        device = devices[addr]
+        flags = True
+        for name in device.operating_flags:
+            if not device.operating_flags[name].is_loaded:
+                flags = False
+                break
+        if flags:
+            for name in device.properties:
+                if not device.properties[name].is_loaded:
+                    flags = False
+                    break
+        if not device.aldb.is_loaded or not flags:
+            await device.async_read_config()
+    await devices.async_save(workdir=config_dir)
+
+
+async def async_setup_platforms(hass, config):
+    """Initiate the connection and services."""
+    for component in INSTEON_COMPONENTS:
+        _LOGGER.info("Setting up insteon platform %s", component)
+        await hass.helpers.discovery.async_load_platform(component, DOMAIN, {}, config)
+        _LOGGER.info("Done setting up up insteon platform %s", component)
+
+    for address in devices:
+        device = devices[address]
+        platforms = get_device_platforms(device)
+        if ON_OFF_EVENTS in platforms:
+            add_on_off_event_device(hass, device)
+
+    _LOGGER.info("Insteon device count: %s", len(devices))
+    register_new_device_callback(hass, config)
+    async_register_services(hass)
+
+    for address in devices:
+        await devices[address].async_status()
```

Comment:
Await inside loop. Can we do it concurrently?

Reply:
I tried that during development but the underlying protocol is a serial protocol with an async wrapper. 90% of the time concurrent will work but some percent of the device status responses get lost whenever I try to make it concurrent. I have not given up on that yet but I am choosing reliability over speed here.

Reply:
Ok. Please write a comment in all places where we need to keep await inside loop. It might be cleaned up otherwise, when we do code maintenance.